### PR TITLE
Trifecta Keyboard

### PR DESCRIPTION
**Pull request by 2Moons-JP**
Trifecta Keyboard

## Description

An Alice style board with arrow keys. Aluminum and Polycarb versions with options for various top plates. 

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [X] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist

- [X] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [X] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [X] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [X] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: keyboards/basekeys/trifecta/config.h [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#pragma once
+
+#include "config_common.h"
+
+/* USB Device descriptor parameter */
+#define VENDOR_ID       0x04D8
+#define PRODUCT_ID      0xEAF3
```
Comment (drashna) [inline keyboards/basekeys/trifecta/config.h:23]: Could you use a different VID here?  This belongs to Microchip. 
https://www.the-sz.com/products/usbid/index.php?v=0x04D8&p=&n=

For instance: 
```suggestion
#define VENDOR_ID       0x6541
#define PRODUCT_ID      0x4277
```

File: keyboards/basekeys/trifecta/config.h [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+
+#define ENCODERS_PAD_A { D4, D6, D7 }
+#define ENCODERS_PAD_B { B4, D3, D5 }
+#define ENCODER_RESOLUTION 4
+
+/* COL2ROW, ROW2COL*/
+#define DIODE_DIRECTION COL2ROW
+
+// Current limiting.
+#define USB_MAX_POWER_CONSUMPTION 500     // Limit device max power consumption.
```
Comment (drashna) [inline keyboards/basekeys/trifecta/config.h:47]: This is the default value and doesn't need to be defined. 
```suggestion
```

**Review by drashna — CHANGES_REQUESTED** [review]

File: keyboards/basekeys/trifecta/config.h (hunk 1 above)
Comment (2Moons-JP) [reply keyboards/basekeys/trifecta/config.h:23]: I received this valid vid and pid from Microchips USB sublicensing program. You can view the application process here:
https://www.microchip.com/usblicensing

File: keyboards/basekeys/trifecta/config.h [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+
+#include "config_common.h"
+
+/* USB Device descriptor parameter */
+#define VENDOR_ID       0x04D8
+#define PRODUCT_ID      0xEAF3
+#define DEVICE_VER      0x0001
+#define MANUFACTURER    Swiftrax
+#define PRODUCT         Trifecta
+#define DESCRIPTION     Alice like keyboard with arrow keys and support for rotary encoders
```
Comment (fauxpark) [inline keyboards/basekeys/trifecta/config.h:27]:
```suggestion
```

File: keyboards/basekeys/trifecta/config.h [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+
+/* key matrix pins */
+
+#define MATRIX_ROW_PINS { B0, B7, F7, B1, B6, C6, C7, B5, F6, D2 }
+#define MATRIX_COL_PINS { F0, F1, F4, F5, D1, B2, D0, B3 }
+
+
+#define ENCODERS_PAD_A { D4, D6, D7 }
+#define ENCODERS_PAD_B { B4, D3, D5 }
+#define ENCODER_RESOLUTION 4
```
Comment (fauxpark) [inline keyboards/basekeys/trifecta/config.h:41]:
```suggestion
```

File: keyboards/basekeys/trifecta/keymaps/default/keymap.c [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+    _______, RGB_TOG, RGB_HUD, RGB_SAD,  RGB_VAD, _______, _______, _______, _______, _______, _______, _______,  _______,                   KC_PGUP,         \
+    _______, _______, _______,           _______, _______,          _______,          _______,          _______,                    KC_HOME, KC_PGDN, KC_END  \
+  ),
+
+  [_MEDIA] = LAYOUT(
+     KC_GRV,   KC_F1,   KC_F2,   KC_F3,   KC_F4,   KC_F5,   KC_F6,   KC_F7,   KC_F8,   KC_F9,  KC_F10,  KC_F11,  KC_F12, _______,  _______,     _______,      \
+    _______, _______, _______, _______,    RESET, _______, _______, _______, _______, _______, _______, _______, _______, _______,               _______,     \
+    _______, RGB_MOD, RGB_HUI, RGB_SAI,  RGB_VAI, _______, _______, _______, _______, _______, _______, _______, _______,                         _______,    \
+    _______, RGB_TOG, RGB_HUD, RGB_SAD,  RGB_VAD, _______, _______, _______, _______, _______, _______, _______,  _______,                   KC_PGUP,         \
+    _______, _______, _______,           _______, _______,          _______,          _______,          _______,                    KC_HOME, KC_PGDN, KC_END  \
```
Comment (fauxpark) [inline keyboards/basekeys/trifecta/keymaps/default/keymap.c:54]:
```suggestion
  [_BASE] = LAYOUT(
    KC_ESC,  KC_1,    KC_2,    KC_3,    KC_4,    KC_5,    KC_6,    KC_7,    KC_8,    KC_9,    KC_0,    KC_MINS, KC_EQL,  KC_BSPC,  KC_DEL,   KC_DEL,
    KC_TAB,  KC_Q,    KC_W,    KC_E,    KC_R,    KC_T,             KC_Y,    KC_U,    KC_I,    KC_O,    KC_P,    KC_LBRC, KC_RBRC, KC_BSLS,     KC_END,
    KC_CAPS, KC_A,    KC_S,    KC_D,    KC_F,    KC_G,             KC_H,    KC_J,    KC_K,    KC_L,    KC_SCLN, KC_QUOT, KC_ENT,                 KC_HOME,
    KC_LSFT, KC_Z,    KC_X,    KC_C,    KC_V,    KC_B,        MO(_MEDIA),   KC_N,    KC_M,    KC_COMM, KC_DOT,  KC_SLSH, KC_RSFT,        KC_UP,
    KC_LCTL, KC_LGUI,  KC_LALT,         KC_SPC,                         KC_SPC,           KC_RGUI, KC_RCTL,  MO(_FN),        KC_LEFT, KC_DOWN, KC_RGHT
  ),

  [_FN] = LAYOUT(
     KC_GRV,   KC_F1,   KC_F2,   KC_F3,   KC_F4,   KC_F5,   KC_F6,   KC_F7,   KC_F8,   KC_F9,  KC_F10,  KC_F11,  KC_F12, _______,  _______,     _______,
    _______, _______, _______, _______,  _______, _______, _______, _______, _______, _______, _______, _______, _______, _______,               _______,
    _______, RGB_MOD, RGB_HUI, RGB_SAI,  RGB_VAI, _______, _______, _______, _______, _______, _______, _______, _______,                         _______,
    _______, RGB_TOG, RGB_HUD, RGB_SAD,  RGB_VAD, _______, _______, _______, _______, _______, _______, _______,  _______,                   KC_PGUP,
    _______, _______, _______,           _______, _______,          _______,          _______,          _______,                    KC_HOME, KC_PGDN, KC_END
  ),

  [_MEDIA] = LAYOUT(
     KC_GRV,   KC_F1,   KC_F2,   KC_F3,   KC_F4,   KC_F5,   KC_F6,   KC_F7,   KC_F8,   KC_F9,  KC_F10,  KC_F11,  KC_F12, _______,  _______,     _______,
    _______, _______, _______, _______,    RESET, _______, _______, _______, _______, _______, _______, _______, _______, _______,               _______,
    _______, RGB_MOD, RGB_HUI, RGB_SAI,  RGB_VAI, _______, _______, _______, _______, _______, _______, _______, _______,                         _______,
    _______, RGB_TOG, RGB_HUD, RGB_SAD,  RGB_VAD, _______, _______, _______, _______, _______, _______, _______,  _______,                   KC_PGUP,
    _______, _______, _______,           _______, _______,          _______,          _______,          _______,                    KC_HOME, KC_PGDN, KC_END
```

File: keyboards/basekeys/trifecta/keymaps/via/keymap.c [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+    _______, RGB_TOG, RGB_HUD, RGB_SAD,  RGB_VAD, _______, _______, _______, _______, _______, _______, _______,  _______,                   KC_PGUP,         \
+    _______, _______, _______,           _______, _______,          _______,          _______,          _______,                    KC_HOME, KC_PGDN, KC_END  \
+  ),
+
+  [_MEDIA] = LAYOUT(
+     KC_GRV,   KC_F1,   KC_F2,   KC_F3,   KC_F4,   KC_F5,   KC_F6,   KC_F7,   KC_F8,   KC_F9,  KC_F10,  KC_F11,  KC_F12, _______,  _______,     _______,      \
+    _______, _______, _______, _______,    RESET, _______, _______, _______, _______, _______, _______, _______, _______, _______,               _______,     \
+    _______, RGB_MOD, RGB_HUI, RGB_SAI,  RGB_VAI, _______, _______, _______, _______, _______, _______, _______, _______,                         _______,    \
+    _______, RGB_TOG, RGB_HUD, RGB_SAD,  RGB_VAD, _______, _______, _______, _______, _______, _______, _______,  _______,                   KC_PGUP,         \
+    _______, _______, _______,           _______, _______,          _______,          _______,          _______,                    KC_HOME, KC_PGDN, KC_END  \
```
Comment (fauxpark) [inline keyboards/basekeys/trifecta/keymaps/via/keymap.c:54]:
```suggestion
  [_BASE] = LAYOUT(
    KC_ESC,  RGB_TOG,    KC_2,    KC_3,    KC_4,    KC_5,    KC_6,    KC_7,    KC_8,    KC_9,    KC_0,    KC_MINS, KC_EQL,  KC_BSPC,  KC_DEL,   KC_DEL,
    KC_TAB,  KC_Q,    KC_W,    KC_E,    KC_R,    KC_T,             KC_Y,    KC_U,    KC_I,    KC_O,    KC_P,    KC_LBRC, KC_RBRC, KC_BSLS,     KC_END,
    KC_CAPS, KC_A,    KC_S,    KC_D,    KC_F,    KC_G,             KC_H,    KC_J,    KC_K,    KC_L,    KC_SCLN, KC_QUOT, KC_ENT,                 KC_HOME,
    KC_LSFT, KC_Z,    KC_X,    KC_C,    KC_V,    KC_B,        MO(_MEDIA),   KC_N,    KC_M,    KC_COMM, KC_DOT,  KC_SLSH, KC_RSFT,        KC_UP,
    KC_LCTL, KC_LGUI,  KC_LALT,         KC_SPC,                         KC_SPC,           KC_RGUI, KC_RCTL,  MO(_FN),        KC_LEFT, KC_DOWN, KC_RGHT
  ),

  [_FN] = LAYOUT(
     KC_GRV,   KC_F1,   KC_F2,   KC_F3,   KC_F4,   KC_F5,   KC_F6,   KC_F7,   KC_F8,   KC_F9,  KC_F10,  KC_F11,  KC_F12, _______,  _______,     _______, 
    _______, _______, _______, _______,  _______, _______, _______, _______, _______, _______, _______, _______, _______, _______,               _______,
    _______, RGB_MOD, RGB_HUI, RGB_SAI,  RGB_VAI, _______, _______, _______, _______, _______, _______, _______, _______,                         _______,
    _______, RGB_TOG, RGB_HUD, RGB_SAD,  RGB_VAD, _______, _______, _______, _______, _______, _______, _______,  _______,                   KC_PGUP,
    _______, _______, _______,           _______, _______,          _______,          _______,          _______,                    KC_HOME, KC_PGDN, KC_END
  ),

  [_MEDIA] = LAYOUT(
     KC_GRV,   KC_F1,   KC_F2,   KC_F3,   KC_F4,   KC_F5,   KC_F6,   KC_F7,   KC_F8,   KC_F9,  KC_F10,  KC_F11,  KC_F12, _______,  _______,     _______,
    _______, _______, _______, _______,    RESET, _______, _______, _______, _______, _______, _______, _______, _______, _______,               _______,
    _______, RGB_MOD, RGB_HUI, RGB_SAI,  RGB_VAI, _______, _______, _______, _______, _______, _______, _______, _______,                         _______,
    _______, RGB_TOG, RGB_HUD, RGB_SAD,  RGB_VAD, _______, _______, _______, _______, _______, _______, _______,  _______,                   KC_PGUP,
    _______, _______, _______,           _______, _______,          _______,          _______,          _______,                    KC_HOME, KC_PGDN, KC_END
```

File: keyboards/basekeys/trifecta/rules.mk [hunk 7]
```diff
@@ -0,0 +1,27 @@
+# SRC = matrix.c
+# SRC += config_led.c
+
```
Comment (fauxpark) [inline keyboards/basekeys/trifecta/rules.mk:3]:
```suggestion
```

File: keyboards/basekeys/trifecta/rules.mk [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+# Do not enable SLEEP_LED_ENABLE. it uses the same timer as BACKLIGHT_ENABLE
+SLEEP_LED_ENABLE = no       # Breathing sleep LED during USB suspend
+# if this doesn't work, see here: https://github.com/tmk/tmk_keyboard/wiki/FAQ#nkro-doesnt-work
+NKRO_ENABLE = no            # USB Nkey Rollover
+BACKLIGHT_ENABLE = no       # Enable keyboard backlight functionality on B7 by default
+RGBLIGHT_ENABLE = yes      # Enable keyboard RGB underglow
+UNICODE_ENABLE = no         # Unicode
+AUDIO_ENABLE = no           # Audio output on port C6
+ENCODER_ENABLE = yes
+# RGB_MATRIX_ENABLE = WS2812
```
Comment (fauxpark) [inline keyboards/basekeys/trifecta/rules.mk:27]:
```suggestion
BOOTMAGIC_ENABLE = lite     # Virtual DIP switch configuration
MOUSEKEY_ENABLE = no        # Mouse keys
EXTRAKEY_ENABLE = yes       # Audio control and System control
CONSOLE_ENABLE = yes        # Console for debug
COMMAND_ENABLE = yes        # Commands for debug and configuration
# Do not enable SLEEP_LED_ENABLE. it uses the same timer as BACKLIGHT_ENABLE
SLEEP_LED_ENABLE = no       # Breathing sleep LED during USB suspend
# if this doesn't work, see here: https://github.com/tmk/tmk_keyboard/wiki/FAQ#nkro-doesnt-work
NKRO_ENABLE = no            # USB Nkey Rollover
BACKLIGHT_ENABLE = no       # Enable keyboard backlight functionality
RGBLIGHT_ENABLE = yes       # Enable keyboard RGB underglow
BLUETOOTH_ENABLE = no       # Enable Bluetooth
AUDIO_ENABLE = no           # Audio output

ENCODER_ENABLE = yes
```

**Review by fauxpark — CHANGES_REQUESTED** [review]

File: keyboards/basekeys/trifecta/rules.mk (hunk 8 above)
Comment (2Moons-JP) [reply keyboards/basekeys/trifecta/rules.mk]: I committed these changes manually.

**Review by drashna — APPROVED** [review]

**Review by fauxpark — APPROVED** [review]

File: keyboards/basekeys/trifecta/readme.md [hunk 9]
```diff
@@ -0,0 +1,17 @@
+# Trifecta
+
+![Trifecta](https://imgur.com/tlIwQow)
```
Comment (noroadsleft) [inline keyboards/basekeys/trifecta/readme.md:3]:
```suggestion
![Trifecta](https://i.imgur.com/tlIwQowl.jpeg)
```

File: keyboards/basekeys/trifecta/readme.md [hunk 10]
```diff
@@ -0,0 +1,17 @@
...
+
+An Alice style board with arrowkeys, based on the Fate Arisu layout.
+
+* Keyboard Maintainer: [swiftrax](https://github.com/swiftrax)  
+* Hardware Supported: Trifecta PCB, ATMEGA32U4-AU
+* Hardware Availability: Hardware Availability: [Website](https://mykeyboard.eu/catalogue/)
```
Comment (noroadsleft) [inline keyboards/basekeys/trifecta/readme.md:9]:
```suggestion
* Hardware Availability: [Website](https://mykeyboard.eu/catalogue/)
```

**Review by noroadsleft — CHANGES_REQUESTED** [review]
Also, you have Tap Dance enabled for both keymaps, but then the key assignment don't use the Tap Dance  keycodes at all. You could remove the Tap Dance code and not turn the feature on, and it would save a bit of space.

**Comment by 2Moons-JP** [comment]
> Also, you have Tap Dance enabled for both keymaps, but then the key assignment don't use the Tap Dance keycodes at all. You could remove the Tap Dance code and not turn the feature on, and it would save a bit of space.

I commited these changes. Thank you.

File: keyboards/basekeys/trifecta/keymaps/default/keymap.c [hunk 11]
```diff
@@ -0,0 +1,78 @@
...
+        if (clockwise) {
+            tap_code(KC_VOLU);
+        } else {
+            tap_code(KC_VOLD);
+        }
+    } else if (index == 1) { /* Second encoder */
+        if (clockwise) {
+            tap_code16(KC_MNXT);
+        } else {
+            tap_code16(KC_MPRV);
```
Comment (fauxpark) [inline keyboards/basekeys/trifecta/keymaps/default/keymap.c:69]:
```suggestion
            tap_code(KC_MNXT);
        } else {
            tap_code(KC_MPRV);
```

**Review by fauxpark — CHANGES_REQUESTED** [review]

**Review by fauxpark — APPROVED** [review]

File: keyboards/basekeys/trifecta/keymaps/default/keymap.c [hunk 12]
```diff
@@ -0,0 +1,78 @@
...
+    _FN,
+    _MEDIA };
+
+// Tap Dance Declarations
+enum {
+    TD_SCAPS = 0
+};
+
+#define SCAPS_LG TD(TD_SCAPS)
+
```
Comment (noroadsleft) [inline keyboards/basekeys/trifecta/keymaps/default/keymap.c:31]:
```suggestion
```

File: keyboards/basekeys/trifecta/keymaps/via/keymap.c [hunk 13]
```diff
@@ -0,0 +1,78 @@
...
+    _FN,
+    _MEDIA };
+
+// Tap Dance Declarations
+enum {
+    TD_SCAPS = 0
+};
+
+#define SCAPS_LG TD(TD_SCAPS)
+
```
Comment (noroadsleft) [inline keyboards/basekeys/trifecta/keymaps/via/keymap.c:31]:
```suggestion
```

**Review by noroadsleft — CHANGES_REQUESTED** [review]

**Review by noroadsleft — APPROVED** [review]

**Comment by noroadsleft** [comment]
Thanks!

**Pull request merged**